### PR TITLE
Add readme

### LINKED ISSUE #11: Add project readme
Add project `README.md` with description what is this library is and how it works.

Also, would be nice to add `CHANGELOG.md`

### PR DESCRIPTION
**Pull request by Tapchicoma**
Close #11

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,208 @@
...
+    data class Deduct(val value: Int) : CalculatorAction()
+    data class Multiply(val value: Int) : CalculatorAction()
+    data class Divide(val value: Int) : CalculatorAction()
+}
+```
+
+With CoRedux you can create a store that will be a single source of current Calculator state 
+and will update it on each new incoming `CalculatorAction`:
+```kotlin
+val store = GlobalScope.createStore<Int, CalculatorAction>(
```
Comment (sockeqwe) [inline README.md]: Should we really add a `GlobalScope` in a README? Likelyhood that people just copy & paste it is high

File: README.md [hunk 2]
```diff
@@ -1,4 +1,208 @@
...
+            is CalculatorAction.Add -> currentState + newAction.value
+            is CalculatorAction.Deduct -> currentState - newAction.value
+            is CalculatorAction.Multiply -> currentState * newAction.value
+            is CalculatorAction.Divide -> currentState / newAction.value
+        }
+    }
+)
+```
+
+Where `reducer` is special function that takes current state, new action that wants to
```
Comment (sockeqwe) [inline README.md]: What do you think about the following:

... is **a** special function responsible for managing state. A `reducer` specify how the state changes in response to actions sent to the store. Remember that actions only describe what happened, but don't describe how the application's state changes. That is the job of the `reducer`.

File: README.md [hunk 3]
```diff
@@ -1,4 +1,208 @@
...
+modify the state (_intention_) and returns result of state modification, that becomes new current state.
+
+To start receiving current state updates, Calculator UI implementation needs to subscribe to `store` instance:
+```kotlin
+store.subscribe { state -> updateUI(state) }
+```
+By default, when first subscription happens, `store` will emit `initialState` to it.
```
Comment (sockeqwe) [inline README.md]: By default means to me that it can be changed i.e. it can be disable to behave like that,  which can't be right?

File: README.md [hunk 4]
```diff
@@ -1,4 +1,208 @@
...
+CoRedux is a predictable state container, that is using same approach as [Redux](https://redux.js.org/) with
+ability to have additional side effects. 
+
+Implementation is based on [Kotlin](https://kotlinlang.org/) [coroutines](https://github.com/Kotlin/kotlinx.coroutines)
+and inspired by both [RxRedux](https://github.com/freeletics/rxredux) library 
+and [Coroutines in practice by Roman Elizarov](https://www.youtube.com/watch?v=a3agLJQ6vt8) KotlinConf2018 talk.
+
+### Basic concept
+
+Imagine, that we need to develop Calculator app that allows to do only following simple actions:
```
Comment (sockeqwe) [inline README.md]: I would introduce the idea of `Action` first, otherwise it's hard to understand `reducer` later

File: README.md [hunk 5]
```diff
@@ -1,4 +1,208 @@
...
+            is CalculatorAction.Divide -> currentState / newAction.value
+        }
+    }
+)
+```
+
+Where `reducer` is special function that takes current state, new action that wants to
+modify the state (_intention_) and returns result of state modification, that becomes new current state.
+
+To start receiving current state updates, Calculator UI implementation needs to subscribe to `store` instance:
```
Comment (sockeqwe) [inline README.md]: Could be simplified: Maybe something like 

> Finally you have to subscribe to get the state updates over time

File: README.md [hunk 6]
```diff
@@ -1,4 +1,208 @@
...
+// Current state will become 10
+store.dispatch(CalculatorAction.Deduct(1)) 
+// Current state will become 9
+store.dispatch(CalculatorAction.Add(1)) 
+// Current state will become 10
+store.dispatch(CalculatorAction.Divide(10)) 
+// Current state will become 1
+```
+
+All actions will be processed in a serialized order
```
Comment (sockeqwe) [inline README.md]: remove `a` ... just `in serialized order`

File: README.md [hunk 7]
```diff
@@ -1,4 +1,208 @@
...
+store.dispatch(CalculatorAction.Divide(10)) 
+// Current state will become 1
+```
+
+All actions will be processed in a serialized order 
+and on each incoming action `reducer` function will be called to update current state.
+
+### Side effects
+
+Side effect is an interface that has one function - it receives incoming
```
Comment (sockeqwe) [inline README.md]: an interface **with** exactly one function - ...

File: README.md [hunk 8]
```diff
@@ -1,4 +1,208 @@
...
+
+Side effect is an interface that has one function - it receives incoming 
+actions, can get current store state at any time and may emit outgoing actions.
+**So basically it is actions in and actions out.** 
+
+**Actions in** to side effects are emitted by store after calling `reducer` function. 
+
+**Actions out** are consumed back again by store and trigger calling `reducer` function.
+
+Basic usage for side effects is to perform some job on incoming action, that may take time (I/O operation)
```
Comment (sockeqwe) [inline README.md]: > Side effects are used to perform additional Job as a reaction on a certain `Action` for example making an http request, I/O operations, writing to database and so on. Since they run in a `Job` they can run async. Each Job is scoped to the context of the `store`.

File: README.md [hunk 9]
```diff
@@ -1,4 +1,208 @@
...
+                        output.send(outputAction)
+                    }
+                }
+            }
+        }
+    }
+}
+```
+
+Now we need to modify how store is created by adding this `sideEffect`:
```
Comment (sockeqwe) [inline README.md]: You must register you SideEffect inside `createStore( sideEffects = listOf(...)`

File: README.md [hunk 10]
```diff
@@ -1,4 +1,208 @@
...
+```kotlin
+val sideEffect = object : SideEffect<Int, CalculatorAction> {
+    override val name: String = "network logger"
+
+    override fun CoroutineScope.start(
+        input: ReceiveChannel<CalculatorAction>,
+        stateAccessor: StateAccessor<Int>,
+        output: SendChannel<CalculatorAction>,
+        logger: SideEffectLogger
+    ): Job = launch(context = CoroutineName(name)) {
```
Comment (sockeqwe) [inline README.md:145]: I'm not sure if this is the best example how you would use a side effect as it focus more on do something if the state == 9. While ok, I think it makes it harder to understand that side effects are meant to react on `Actions` or trigger additional work on a certain `Action`

File: README.md [hunk 11]
```diff
@@ -1,4 +1,208 @@
...
+storeWithSideEffect.subscribe { state -> updateUI(state) }
+storeWithSideEffect.dispatch(CalculatorAction.Add(9)) 
+// Current state will become 9
+storeWithSideEffect.dispatch(CalculatorAction.Add(1)) 
+// Current state will become 10 and network request will happen
+```
+
+Library provides useful simplified versions of `SideEffect` interface:
+- `SimpleSideEffect` that produces on one input action either one output action or no action.
+- `CancellableSideEffect` that cancels currently executed `Job` and starts a new one.
```
Comment (sockeqwe) [inline README.md]: starts a new one if the same type of `Action` is dispatched to the store.

File: README.md [hunk 12]
```diff
@@ -1,4 +1,208 @@
...
+network request from side effect will be performed:
+```kotlin
+storeWithSideEffect.subscribe { state -> updateUI(state) }
+storeWithSideEffect.dispatch(CalculatorAction.Add(9)) 
+// Current state will become 9
+storeWithSideEffect.dispatch(CalculatorAction.Add(1)) 
+// Current state will become 10 and network request will happen
+```
+
+Library provides useful simplified versions of `SideEffect` interface:
```
Comment (sockeqwe) [inline README.md]: CoRedux  also provides two simplified implementations  of `SideEffect` that you might find useful:

File: README.md [hunk 13]
```diff
@@ -1,4 +1,208 @@
...
+- `SimpleSideEffect` that produces on one input action either one output action or no action.
+- `CancellableSideEffect` that cancels currently executed `Job` and starts a new one.
+
+### Implementation details
+
+By default, CoRedux starts main reducer coroutine ("manager"), when `StateReceiver` subscribes to `CoreduxStore` instance.
```
Comment (sockeqwe) [inline README.md]: I would remove the `By default` as again, is it possible to change that? You just can change the default coroutine the `reducer` runs, right? but you can't change the overall behavior that it runs on a "reducer coroutine"

File: README.md [hunk 14]
```diff
@@ -1,4 +1,208 @@
...
+Reducer coroutine itself starts all side effects coroutines ("workers"), sets current state to initial state
+and immediately updates with it `StateReceiver`. Then reducer coroutine starts listening
```
Comment (sockeqwe) [inline README.md]: I would say 

> and immediatley pushed the intial state to the `StateReciever`

File: README.md [hunk 15]
```diff
@@ -1,4 +1,208 @@
...
+for a new action from actions channel.
```
Comment (sockeqwe) [inline README.md]: I would remove `from actions channel`. Action channel seems to me like an implementation detail others don't have to know about. I would just say 

>  Then reducer coroutine starts listening for actions.

File: README.md [hunk 16]
```diff
@@ -1,4 +1,208 @@
...
+- `CancellableSideEffect` that cancels currently executed `Job` and starts a new one.
+
+### Implementation details
+
+By default, CoRedux starts main reducer coroutine ("manager"), when `StateReceiver` subscribes to `CoreduxStore` instance.
+Reducer coroutine itself starts all side effects coroutines ("workers"), sets current state to initial state
+and immediately updates with it `StateReceiver`. Then reducer coroutine starts listening
+for a new action from actions channel. 
+
+When a new action received, reducer coroutine triggers reducer function, that updates current state inside coroutine,
```
Comment (sockeqwe) [inline README.md]: > new action  **is**  received.

> that updates current state inside coroutine
I don't understand that? It changes state of coroutine or do you mean the state of the whole store?

File: README.md [hunk 17]
```diff
@@ -1,4 +1,208 @@
...
+sends updated current state to `StateReceiver`, and then sends input action to all side effects, using broadcast channel.
+
+If side effect emits new action based on input action and current state - 
+this action are added as a last one to reducer coroutine input actions channel and
+will be processed after all older actions will go through reducer function.
```
Comment (sockeqwe) [inline README.md]: Nice explanation 👍

File: README.md [hunk 18]
```diff
@@ -1,4 +1,208 @@
...
+
+### Logging
+
+`createStore` function has as a parameter list of `LogSink`s - provide
```
Comment (sockeqwe) [inline README.md]: I would just say 

> `createStore()` 

instead of
 
> `createStore` function


also instead of 

> provide 

I would say 

> add a `LogSink, if you want to log events`

File: README.md [hunk 19]
```diff
@@ -1,4 +1,208 @@
...
+any `LogSink` implementations, if you want to start receiving log events from the store.
+
+`createStore` and reducer coroutine will automatically produce log events,
+while `SideEffect` implementations should send it themselfs using provided `logger`.
```
Comment (sockeqwe) [inline README.md]: I don't get it ... what means 

> should send it themselfs using provided `logger`.

you mean SideEffects can add log messages by using the logger passed as parameter, right?

I would rephrase this ...

File: README.md (hunk 1 above)
Comment (Tapchicoma) [reply README.md]: Yeah, good point :+1:

File: README.md (hunk 3 above)
Comment (Tapchicoma) [reply README.md]: It can be specifically configured by passing `launchMode`:
```kotlin
val store = scope.createStore(
    ...,
    launchMode = CoroutineStart.DEFAULT,
    ....
)
```

Default value is `CoroutineStart.LAZY`

File: README.md (hunk 10 above)
Comment (Tapchicoma) [reply README.md:145]: It is hard to imagine any side effect that is required for calculator. While it is artificial - I wanted to introduce filtering by both `stateAccessor()` and `action`. I may update the condition to `stateAccessor() >= 0` :thinking:

File: README.md (hunk 13 above)
Comment (Tapchicoma) [reply README.md]: > is it possible to change that

Yes, I've answered about it in some previous reply.

User can define whether reducer coroutine start immediately (so it is possible to dispatch action and update the state) or when first `StateReceiver` subscribes.

File: README.md (hunk 15 above)
Comment (Tapchicoma) [reply README.md]: Though implementation details, kind of, exposed in `SideEffect` interface :thinking: 

What about "from incoming actions queue"?

File: README.md (hunk 16 above)
Comment (Tapchicoma) [reply README.md]: > I don't understand that? It changes state of coroutine or do you mean the state of the whole store?

It changes the current state field inside reducer coroutine.

File: README.md (hunk 19 above)
Comment (Tapchicoma) [reply README.md]: > you mean SideEffects can add log messages by using the logger passed as parameter, right?

Yes, as `SideEffect` is an interface, implementations should add logs by themselfs.

`SimpleSideEffect` and `CancellableSideEffect` adds some logs, but not all. Also it is possible to send custom side effect log message.

File: README.md (hunk 3 above)
Comment (sockeqwe) [reply README.md]: My point is, you can't disbale emitting the `initial state`, right?

> By default, when first subscription happens, `store` will emit `initialState` to it.

This means to me I somehow can configure the `store` not to submit the `initialState`

File: README.md (hunk 15 above)
Comment (sockeqwe) [reply README.md]: does it make sense to explain the action channel in the side effect section?

File: README.md (hunk 16 above)
Comment (sockeqwe) [reply README.md]: For me as reader it doesn't add value, it rather confuses me I guess ... 

Up to you  ...

File: README.md (hunk 19 above)
Comment (sockeqwe) [reply README.md]: Got you, I would rephrase this sentence then

File: README.md (hunk 3 above)
Comment (Tapchicoma) [reply README.md]: Yes, `initialState` may not be dispatched if `launchMode = CoroutineStart.DEFAULT`, as there will be no `StateReceivers` at that moment, though `currentState` will be updated to `initialState`

Comment (Tapchicoma) [reply README.md]: What about such phrasing:

With defaults (`launchMode = CoroutineStart.LAZY`), on first subscription, `store` will emit `initialState` to it.

File: README.md (hunk 15 above)
Comment (Tapchicoma) [reply README.md]: This "Implementation details" section, so I would not explain what is channel in side effects section.

Though I've added a little more explanation here.

File: README.md (hunk 16 above)
Comment (Tapchicoma) [reply README.md]: I've reworked "Implementation details" section.

File: README.md (hunk 19 above)
Comment (Tapchicoma) [reply README.md]: Rephrased.

File: README.md (hunk 10 above)
Comment (Tapchicoma) [reply README.md:145]: Updated.

File: README.md (hunk 13 above)
Comment (Tapchicoma) [reply README.md]: Rephrased it.

**Comment by sockeqwe** [comment]
All the code snippets are missing how to subscribe with `StateReceiver`. I would add this at least add it to the calculater code snipped. Also `StateReceiver` is mentioned in the implementation details without any further introduction what it is ...

Also I think a kotlin dsl to create the store would be really really nice and a great selling point

**Comment by Tapchicoma** [comment]
@sockeqwe actually there is a code snippet about it, but, you are right, what is `StateReceiver` itself is missed description. Will add one.

Regarding DSL - opened an issue: #33. Though adding it is definitely not the goal of initial release.

File: README.md [hunk 20]
```diff
@@ -1,4 +1,223 @@
 # CoRedux
-Experimental port of [RxRedux](https://github.com/freeletics/rxredux) to Kotlin [coroutines](https://github.com/Kotlin/kotlinx.coroutines).
 
-More info to follow...
\ No newline at end of file
+Opinionated [Redux](https://redux.js.org/) implementation using [Kotlin](https://kotlinlang.org/) [coroutines](https://github.com/Kotlin/kotlinx.coroutines) inspired by [RxRedux](https://github.com/freeletics/rxredux).
+
+## Table of content
```
Comment (sockeqwe) [inline README.md:5]: Nice structure 👍

File: README.md [hunk 21]
```diff
@@ -1,4 +1,223 @@
...
+  - [Gradle](#gradle)
+  - [Additional artifacts](#additional-artifacts)
+- [What is CoRedux](#what-is-coredux)
+  - [Basic concept](#basic-concept)
+  - [Side effects](#side-effects)
+  - [Implementation details](#implementation-details)
+  - [Logging](#logging)
+- [Sample app](#sample-app)
+
+## How to use
```
Comment (sockeqwe) [inline README.md]: How to use is fine, but I would probably name it "Setup" or "Getting started". 

I, personally, associate "How to use" with how to create a redux store and how to use in inside my app, not how to setup the gradle dependencies.

File: README.md [hunk 22]
```diff
@@ -1,4 +1,223 @@
...
+Where `reducer` is a special function responsible for managing state. A `reducer` specify
+how the state changes in response to actions sent to the store.
+Remember that actions only describe what happened, but don't describe how the application's state changes.
+That is the job of the `reducer`.
+
+Finally you have to subscribe `StateReceiver` to the `store` instance to get the state updates over time:
+```kotlin
+store.subscribe { state -> updateUI(state) }
+```
+Where `StateReceiver` is a function, that will receive each new update of the state inside `store`.
```
Comment (sockeqwe) [inline README.md]: Remove `Where`.  Also consider rephrasing it. Example:

> `StateReceiver` is a function that is called whenever the state of the store has been changed. You can think of it as a listener of store's state.

File: README.md [hunk 23]
```diff
@@ -1,4 +1,223 @@
...
+More then one `StateReceiver` can subscribe to the same `store` instance.
```
Comment (sockeqwe) [inline README.md]: More th**a**n (not th**e**n)

Comment (sockeqwe) [reply README.md]: Implementation detail related questions:
1. Do I understand this correctly: According to [this code](https://github.com/freeletics/CoRedux/blob/master/library/core/src/main/kotlin/com/freeletics/coredux/ReduxStore.kt#L150) new subscribers don't get the current state of the store immediately, right?
 For example, in the sample calculator app the calculator store has already executed two actions  `Add(2)`, `Add(3)`. So we are no longer in the initial state `0` but the state is `5` (the result of the two Add actions). What happens if a new subscriber starts subscribing now. Does the subscriber gets `5` immediately or have to wait for the next Action to go through the store to update the state to finally notify all `StateReceivers` incl. the new one? In other words, does the CoRedux store behaves similar to Rx's BehaviorSubject for new subscribers?

2. How does unsubscribing works? There is a `store.unsubscribe(StateReceiver)` but at the same time the whole store is scoped to the CoroutineScope.  What happens to the subscribes if a CoroutineScope gets closed?

I think both questions are a highlighting design flaws.

Does it make sense to create a new github issue for each of them?

File: README.md [hunk 24]
```diff
@@ -1,4 +1,223 @@
...
+
+With default parameters (`launchMode = CoroutineStart.LAZY`), on first subscription, `store` will emit `initialState` to it.
```
Comment (sockeqwe) [inline README.md]: I still find this sentence confusing: It implies somehow to me that only if I set `launchMode = CoroutineStart.LAZY` the initial state will be emitted and if you use another launchMode it will not emit the initial state. So there are two things:
1.  In general I find this sentence hard to read and understand without looking at the implementation details of how store is implemented. 
2. I think you mix up 2 concepts here in one sentence. One is when CoRedux store starts to do the work (on first subscription if you use `launchMode = CoroutineStart.LAZY` or for example immediately with other launch modes like `CoroutineStart.DEFAULT`) and the other concept is what will a `StateReceiver` receive on subscribing. Maybe it's worth to split this two concepts into two different sentences (maybe even a dedicates section about launchMode). If I understood it correctly the fact that `CoroutineStart.LAZY` is emitting initalState to state receiver is mostly just because otherwise (i.e. with `CoroutineStart.DEFAULT`) the state initial state will be emitted to an empty list of `StateReceiver` because `store.subscribe()` didn't happen fast enough. Also, could it be possible that even with `CoroutineStart.DEFAULT` the initial state could be emitted to StateReceiver i.e. if somehow the reducer is having a `delay(10_000)` or if `initialState = 0.also { delay(10_000) }` and store.subscriber() is called on a different coroutinescope or thread? Again, I think we are mixing here up two things and we should make it very clear what new subscribers get emitted on subscription. I would advocate for immediately pass the emit the current state no matter if you are the first subscriber or if you subscribed much much later after some actions have already triggered some actions. Again, maybe the current design is not the best (also related to my questions one line above)?

File: README.md [hunk 25]
```diff
@@ -1,4 +1,223 @@
...
+store.dispatch(CalculatorAction.Deduct(1))
+// Current state will become 9
+store.dispatch(CalculatorAction.Add(1))
+// Current state will become 10
+store.dispatch(CalculatorAction.Divide(10))
+// Current state will become 1
+```
+
+All actions will be processed in serialized order
+and on each incoming action `reducer` function will be called to update current state.
```
Comment (sockeqwe) [inline README.md]: is called instead of will be called

File: README.md [hunk 26]
```diff
@@ -1,4 +1,223 @@
...
+
+Side effects are used to perform additional `Job` as a reaction on a certain action, for example,
+making an HTTP request, I/O operations, writing to database and so on. Since they run in a `Job` they can run async.
+Each `Job` is scoped to the context of the `store`.
+
+Let's add an side effect that makes a network request each time when action is `CalculatorAction.Add`
+and current state is `9`. If server responds with http code `200` - side effect should emit `CalculatorAction.Deduct(1)`.
+```kotlin
+val sideEffect = object : SideEffect<Int, CalculatorAction> {
+    override val name: String = "network logger"
```
Comment (sockeqwe) [inline README.md:138]: Nitpicking: 

If I understood it corretly `SideEffect.name` is only used for logging, right? What if I don't use the logging capabilities at all, then I still have to  set a name on my side effect, right? Maybe not a big deal but seems to me like  a little design flaw. Maybe the logger can determine the name on demand (i.e. how Jake Wharton's Timber determines the calling function name) or completely let it to the developer using `logger` to include this "side effect name" information in the log message.

File: README.md [hunk 27]
```diff
@@ -1,4 +1,223 @@
...
+```kotlin
+storeWithSideEffect.subscribe { state -> updateUI(state) }
+storeWithSideEffect.dispatch(CalculatorAction.Deduct(1))
+// Current state will become -1
+storeWithSideEffect.dispatch(CalculatorAction.Add(11))
+// Current state will become 10 and network request will happen
+```
+
+`CoRedux` also provides two simplified implementations of `SideEffect` that you might find useful:
+- `SimpleSideEffect` that produces on one input action either one output action or no action.
```
Comment (sockeqwe) [inline README.md]: I think it would be nice to have a little code snipped for this one as well as I think many people making http requests will find this one useful. Maybe even put `SimpleSideEffect` with code snipped before explaining the general purpose  `object:SideEffect` implementation. What do you think?

File: README.md [hunk 28]
```diff
@@ -1,4 +1,223 @@
...
+- `CancellableSideEffect` that cancels previously running `Job` and starts a new one if the same type of action is dispatched to the store.
+
+### Implementation details
+
+With default values in `createStore()`, CoRedux starts main reducer coroutine ("manager"), when first `StateReceiver` subscribes to `CoreduxStore` instance.
```
Comment (sockeqwe) [inline README.md]: Isn't it repeating  / very similar what you already said before in 

> With default parameters (`launchMode = CoroutineStart.LAZY`), on first subscription, `store` will emit `initialState` to it.

Again, maybe the whole thing of launchMode, StateReceiver and "manager" coroutine should be simplified and put into one place?

Could a simplification for this particular sentence 

> With default values in `createStore()`, CoRedux starts main reducer coroutine ("manager"), when first `StateReceiver` subscribes to `CoreduxStore` instance.

be something like :

> Internally the reducer runs on its own coroutine. This ensures that only one action is processed by the reducer at one time and avoids concurrency problems from the very beginning.

File: README.md [hunk 29]
```diff
@@ -1,4 +1,223 @@
...
+"Manager" coroutine is a single source of current state that is defined in local scope of coroutine - this allows to prevent any concurrency problems.
+Then "manager" coroutine itself sequentially:
```
Comment (sockeqwe) [inline README.md]: There is a verb missing. Maybe

> Then "manager" coroutine itself **is**  sequentially:

I personally would also replace "Then" with "Furthermore".

File: README.md [hunk 30]
```diff
@@ -1,4 +1,223 @@
...
+- starts all side effects coroutines ("workers"), that listens for incoming actions
+- sets current state to initial state
+- immediately pushes the initial state to the `StateReciever`
+- starts listening for a new action from actions `ReceiveChannel`,
+that are populated by either from `dispatch(new_action)` call or from side effects output actions `SendChannel`.
+
+On each new action from `ReceiveChannel` "manager" coroutine sequentially:
```
Comment (sockeqwe) [inline README.md]: Again, a verb is missing ...

File: README.md [hunk 31]
```diff
@@ -1,4 +1,223 @@
...
+
+### Implementation details
+
+With default values in `createStore()`, CoRedux starts main reducer coroutine ("manager"), when first `StateReceiver` subscribes to `CoreduxStore` instance.
+"Manager" coroutine is a single source of current state that is defined in local scope of coroutine - this allows to prevent any concurrency problems.
+Then "manager" coroutine itself sequentially:
+- starts all side effects coroutines ("workers"), that listens for incoming actions
+- sets current state to initial state
+- immediately pushes the initial state to the `StateReciever`
+- starts listening for a new action from actions `ReceiveChannel`,
```
Comment (sockeqwe) [inline README.md]: Is it really needed to mention `ReceiverChannel` here? 

For me as a reader it is more confusing then helping. I think it exposes too much implementation details. Imho it is enough to write something like:

> starts listening for new actions emitted through `store.dispatch(action)` and from `SideEffects`

If that is not enough for the reader, I guess as a reader I would expect to get a link from `store.dispatch()` to `ReceiveChannel` at least.

File: README.md [hunk 32]
```diff
@@ -1,4 +1,223 @@
...
+- `CancellableSideEffect` that cancels previously running `Job` and starts a new one if the same type of action is dispatched to the store.
+
+### Implementation details
+
+With default values in `createStore()`, CoRedux starts main reducer coroutine ("manager"), when first `StateReceiver` subscribes to `CoreduxStore` instance.
+"Manager" coroutine is a single source of current state that is defined in local scope of coroutine - this allows to prevent any concurrency problems.
+Then "manager" coroutine itself sequentially:
+- starts all side effects coroutines ("workers"), that listens for incoming actions
+- sets current state to initial state
+- immediately pushes the initial state to the `StateReciever`
```
Comment (sockeqwe) [inline README.md]: Again, somehow related to 

> With default parameters (launchMode = CoroutineStart.LAZY), on first subscription, store will emit initialState to it.

It might confuses the reader or at least make him / her wonder if it is the same thing or not

File: README.md [hunk 33]
```diff
@@ -1,4 +1,223 @@
...
+- starts listening for a new action from actions `ReceiveChannel`,
+that are populated by either from `dispatch(new_action)` call or from side effects output actions `SendChannel`.
+
+On each new action from `ReceiveChannel` "manager" coroutine sequentially:
+- triggers `reducer()` to update current state inside "manager" coroutine local scope
+- sends updated current state to all `StateReceiver`s
+- sends input action to all side effects, using `BroadcastChannel`
```
Comment (sockeqwe) [inline README.md]: > using an **internal** `BroadcastChannel`

File: README.md [hunk 34]
```diff
@@ -1,4 +1,223 @@
...
+
+If side effect emits new action based on input action and current state -
```
Comment (sockeqwe) [inline README.md]: I would simplify this again (also remove some more implementation details like ReceiverChannel):

> If **a** side effect emtis **a** new action this action is added at the end of the internal queue of actions waiting to be dispatched  to reducer and to other side effects.

File: README.md [hunk 35]
```diff
@@ -1,4 +1,223 @@
...
+this action is added as a last one to "manager" coroutine input actions `ReceiveChannel` and
+will be processed after all older actions will go through "manager" coroutine input action loop.
+
+### Logging
+
+`createStore()` has as a parameter list of `LogSink`s - add `LogSink` implementations, if you want to log events.
```
Comment (sockeqwe) [inline README.md]: needless `,`


Also I think this sentence could be rephrased to something like:

> To log events you have to add a `LogSink` to your store. Actually, you can add multiple `LogSinks` if you want to add multiple type of logging. You do that by passing a list of `LogSink`s  in `createStore( logSinks = listOf(logSink1, logSink2, ... )`

File: README.md [hunk 36]
```diff
@@ -1,4 +1,223 @@
...
+
+`createStore()` and "manager" coroutine will automatically produce log events,
+though `SideEffect` implementations, as library does not have control over it, should send them using provided `logger`.
```
Comment (sockeqwe) [inline README.md]: Incorrect grammar. 

Also, I don't understand what you are trying to say. Either rephrase it or remove it as from my point of view the previous sentence is enough, isn't it?

> `createStore()` has as a parameter list of `LogSink`s - add `LogSink` implementations, if you want to log events.

File: README.md (hunk 25 above)
Comment (sockeqwe) [reply README.md]: Also I think 

> called to update the state

is not correct (and implies that state is mutable). I would change it to

> called to compute new state

File: README.md (hunk 23 above)
Comment (Tapchicoma) [reply README.md]: > According to this code new subscribers don't get the current state of the store immediately, right?

Right, second subscriber will not receive current state on subscribe. Whenever next action will be dispatched, second subscriber will get updated state, so behaviour is similar to Rx `PublishSubject`.

> How does unsubscribing works?

Currently no unsubscribtion happens, when `CoroutineScope` is closed, but on trying to dispatch new action exception will be thrown. I think it is worth to add  a code that will remove all `StateReceiver`s from the store to prevent memory leak. Created an issue for that: https://github.com/freeletics/CoRedux/issues/34. Thank you for this catch :+1:

File: README.md (hunk 24 above)
Comment (Tapchicoma) [reply README.md]: > If I understood it correctly the fact that CoroutineStart.LAZY is emitting initalState to state receiver is mostly just because otherwise (i.e. with CoroutineStart.DEFAULT) the state initial state will be emitted to an empty list of StateReceiver because store.subscribe() didn't happen fast enough.

Yes, in `CoroutineStart.DEFAULT` case, reducer coroutine will immediately start, set current state to `initialState` and will wait for a new actions. No `StateReceiver`s will receive `initialState`, as no is subscribed at that moment.

> Also, could it be possible that even with CoroutineStart.DEFAULT the initial state could be emitted to StateReceiver i.e. if somehow the reducer is having a delay(10_000) or if initialState = 0.also { delay(10_000) } and store.subscriber() is called on a different coroutinescope or thread?

Using `delay(...)` in `reducer` function or `State.also { .. }` is not possible (needs to be in coroutine context/`suspend fun`). I can think of using, possibly instead, `Thread.sleep(...)`, but I even wonder what will happen. I'd say it is edge case.

> I would advocate for immediately pass the emit the current state no matter if you are the first subscriber or if you subscribed much much later after some actions have already triggered some actions.

I would not change current approach:
- with `CoroutineStart.DEFAULT` - you can use "headless" store and only receive state updates by subscribing/unsubscribing to it. Kind of event bus for particular state.
- if user needs "headless" store, but also wants to receive last state - they can use combination of `CoroutineStart.LAZY` with custom wrapper around `Store` interface (possibly worth to include in core artefact), that will store last state and propagate it to new subscribers.

I would try to better formulate this moment.

File: README.md (hunk 26 above)
Comment (Tapchicoma) [reply README.md:138]: > If I understood it corretly SideEffect.name is only used for logging, right?

Correct.

> Maybe the logger can determine the name on demand (i.e. how Jake Wharton's Timber determines the calling function name)

Needs investigation, can be added later as default interface `val name` implementation.

> completely let it to the developer using logger to include this "side effect name" information in the log message.

I would not do that - this will mess with expected log events from the `SideEffect` if developer will use different names.

File: README.md (hunk 32 above)
Comment (Tapchicoma) [reply README.md]: Yes, seems this need clarification - will add.

File: README.md (hunk 31 above)
Comment (Tapchicoma) [reply README.md]: Yeah, your phrasing is better :+1:

File: README.md (hunk 35 above)
Comment (Tapchicoma) [reply README.md]: Nice phrasing :+1:

File: README.md (hunk 36 above)
Comment (Tapchicoma) [reply README.md]: I wanted to say that except `SideEffect` implementation other parts will produce log events automatically.

Comment (Tapchicoma) [reply README.md]: I would rephrase this.

File: README.md (hunk 27 above)
Comment (Tapchicoma) [reply README.md]: Added examples for both.

**Comment by Tapchicoma** [comment]
@sockeqwe I've added a section about `launchMode`. Imho, now README is looking good.

**Review by sockeqwe — APPROVED** [review]
